feat(helm): add ability for --dry-run to do lookup functions

tapaskapadia opened the pull request:
When a helm command is run with the --dry-run flag, it will try to connect to the cluster
to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/master/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**: It is hard to debug the lookup function and currently there is not a good way to test it with any flags. #8137 Stated that the `--dry-run` was fair game to try to implement this logic as long as the `helm template` logic stays the same.

**Special notes for your reviewer**: This is my first PR for the Helm; please let me know if I need to add or change anything. 

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

tapaskapadia committed:
feat(helm): add ability for --dry-run to do lookup functions

When a helm command is run with the --dry-run flag, it will try to connect to the cluster
to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

tapaskapadia commented:
FYI, I believe this test is not testing what it states. I think it will pass regardless of the DryRun field being sent to "true" or "false". https://github.com/helm/helm/blob/master/pkg/action/install_test.go#L245

I have attached a document of a sample of the basic tests I ran. I did not see any good way to positively test a successful lookup without using a live cluster. 

[basicTests.txt](https://github.com/helm/helm/files/6056441/basicTests.txt)

DanielHabenicht commented:
Any updates on this? I would support if something is missing.

iTaybb commented:
I'm also looking forward for this, and wish to help if any help is needed.

rickardp commented:
@tapaskapadia  Can you fix the merge conflicts? Or if you let me, I'm happy to do so. I can help out with testing if needed.

rickardp commented on pkg/action/action.go:
```diff
@@ -120,12 +120,12 @@ func (c *Configuration) renderResources(ch *chart.Chart, values chartutil.Values
-	// wishes and do not connect to the cluster.
-	if !dryRun && c.RESTClientGetter != nil {
+	// A `helm template`  should not talk to the remote cluster. However, commands
+	// with `--dry-run` should be able to try to connect to the cluster.
```
This change is much more in line with user expectations, IMHO. Specifically, the sentence saying "when the user says to dry run, respect the user's wishes and do not connect to the cluster" is not in any way implied by naming convention or any best practices. Typically, dry run an operations means "do not mutate" not "do not read". Are there any documentation to update as well?

tapaskapadia replied:
Good call on updating the documentation: https://helm.sh/docs/chart_template_guide/functions_and_pipelines/#using-the-lookup-function
This page does reference that helm install --dry-run is not supposed to contact the cluster. 
If this is approved this page will also need to be changed: https://github.com/helm/helm-www/blob/2875cecd9b1b8b6290e9fee595e288807f033eb5/content/ko/docs/chart_template_guide/functions_and_pipelines.md

tapaskapadia commented:
> @tapaskapadia Can you fix the merge conflicts? Or if you let me, I'm happy to do so. I can help out with testing if needed.
I'll rebase it to fix the merge conflicts. I may need a little help getting started if unit test cases are required as when I initally took at a stab at this it was not clear to be how to simulating having a real connection.

mumoshu commented:
FWIW, `helm-diff` needs this, and I prefer this over changing `helm template` to have cluster access.

https://github.com/databus23/helm-diff/issues/263

joejulian commented:
I would prefer we not do this in this fashion, but instead do what kubectl does with `--dry-run=<server|client>`. I think an interim change to just default to `--dry-run=client` if `--dry-run` is specified without qualification could be sufficient to keep it from being considered a breaking change, then go ahead and open up the flood gates on server connections with `--dry-run=server`.

Unfortunately, dry-run already broke the client-only contract. I'd still be against changing the comments to make it seem ok to do that further, especially if we're going to want to walk that back in helm 4.

beekhof commented:
I found myself nodding furiously to this comment from @rickardp
> Specifically, the sentence saying "when the user says to dry run, respect the user's wishes and do not connect to the cluster" is not in any way implied by naming convention or any best practices. Typically, dry run an operations means "do not mutate" not "do not read".

That this topic is raised over and over again is surely a sign that dry-run _must_ imply a client-only contract is not what everyone wants, because it makes Helm less useful for some of us, and our lives harder.

@joejulian 's `--dry-run=<server|client>` suggestion seems pretty reasonable compromise. 
* Are there key folks that remain to be convinced before such an approach could be merged?
* Is someone already working on it, or are we waiting for volunteers? (I'd be happy to do so).

ksa-real commented:
+1 on that one. I had to re-implement deployment scripts to do necessary lookups outside of `helm upgrade --dry-run` and then supply retrieved values via a temp file. My use of dry-run is to do a preflight check and `helm diff` with the already deployed release. At the moment not having lookup in -dry-mode generates too much notice and makes this mode too hard to use (without my enhancements).

v-sag commented:
Please support this.

QuentinN42 commented:
Any updates about the needed work to merge it ?
I have some time to pass some commits :)

alexzhc commented:
ugently needed 👍🏻

jgreat commented:
Can I ask a question here? Would this feature be more likely to be approved and merged if it wasn't tied to the --dry-run flag? 

I can't speak for everyone, but what I'm looking for is the ability to do a full test run with connections to the remote cluster so lookups work. These cluster connections would run under the user's approved/provisioned security context, so there doesn't seem to be a security problem or potential to expose something that the user doesn't already have access to via kubectl.

QuentinN42 commented:
Looking at the k8s docs of the `dry-run` :
> The request is still processed as typical request: the fields are defaulted, the object is validated, it goes through the validation admission chain, and through the mutating admission chain, and then the final object is returned to the user as it normally would, without being persisted.
So `kubectl` notion of dry-run is "I do all the stuff as normal, but at the last moment I do not apply changes".

I think `helm upgrade` with a `--dry-run` must do the same, lookups are executed, connection to the cluster is tested, and objects are validated through the k8s API.

If you want to NOT connect to the cluster, what is the difference with a `helm template` ?

joejulian commented:
Looking at kubectl help:

```
    --dry-run='none':
	Must be "none", "server", or "client". If client strategy, only print the object that
	would be sent, without sending it. If server strategy, submit server-side request without
	persisting the resource.
```

I'm suggesting helm would do the same.

mikeshng commented:
+1 on this PR or any other PRs that can enhance the dry-run capability to perform lookup.

QuentinN42 commented:
@rickardp what is needed to merge this ?
@tapaskapadia are you still working on this ?

If needed I can help by creating a new MR with all this changes on the last version of `helm:main`.

gjenkins8 commented:
I second @joejulian said about `--dry-run='none'|'client'|'server'`. The PR here (as it stands) is a change in behavior. 

Currently `helm install --dry-run=true` does not contact the cluster, and to the point of the security advisory/model ([GHSA-q8q8-93cv-v6h8](https://github.com/helm/helm/security/advisories/GHSA-q8q8-93cv-v6h8)), I think would be a breaking change for some users (ie. they could suddenly get secrets printed and logged in a CI system; see e.g. https://github.com/helm/helm/issues/7275)

I put a proposal on the main issue (rather than this implementation PR [here](https://github.com/helm/helm/issues/8137#issuecomment-1365384675)

sandipchitale commented:
What about Capabilities.APIVersions dictionary? Will this also work once lookup will work?

pwillis-eiq commented:
It sounds like there is some consensus (at least from some people aligned with the project) on how to move forward. Preserve the old Helm options and functionality (`--dry-run=true`, `--dry-run=false`), and add new functionality (`--dry-run=client`, `--dry-run=server`, `--dry-run=none`).

I literally don't know how to program in Go, so it would take me an unreasonable amount of time to prep this change. But if anyone on this thread *does* program in Go, feel free to contact me at `peter at mutable dot cc`, and I will work with you to walk through the code, prepare the changes, write docs, test code, etc. It's open source, ya'll... team work makes the dream work :)

tapaskapadia committed:
feat(helm): add ability for --dry-run to do lookup functions

When a helm command is run with the --dry-run flag, it will try to connect to the cluster
if the value is 'server' to be able to render lookup functions.
Closes helm#8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

QuentinN42 commented:
Hey @pwillis-eiq ,
Discussions has moved to this issue :
https://github.com/helm/helm/issues/8137#issuecomment-1365384675
Yup, if you want, some docs needs to be written and maybe some other things to think about.

tapaskapadia committed:
feat(helm): add ability for --dry-run to do lookup functions
When a helm command is run with the --dry-run flag, it will try to connect to the cluster
if the value is 'server' to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

tapaskapadia committed:
feat(helm): add ability for --dry-run to do lookup functions
When a helm command is run with the --dry-run flag, it will try to connect to the cluster
if the value is 'server' to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

tapaskapadia commented:
Hey @QuentinN42 and @pwillis-eiq, 
I spent some time reworking the PR. To accommodate for both the old and new flags, I had to change the flag from bool to string. Let me know if I need to change anything or if there is more needed.

| command       | dryRun Flag Value | expected behavior |
| ------------- | ------------- | ------------- |
| install  | --dry-run=false  | not a dry run  |
| install  | --dry-run=none  | not a dry run  |
| install  | [no flag present]  | not a dry run  |
| install  | --dry-run  | dry run w/ no server interaction  |
| install  | --dry-run=true | dry run w/ no server interaction  |
| install  | --dry-run=false | dry run w/ no server interaction  |
| install  | --dry-run=client | dry run w/ no server interaction  |
| install  | --dry-run=server | dry run with server interaction  |
| upgrade --install  | --dry-run=false  | not a dry run  |
| upgrade --install  | --dry-run=none  | not a dry run  |
| upgrade --install  | [no flag present]  | not a dry run  |
| upgrade --install  | --dry-run  | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=true | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=false | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=client | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=server | dry run with server interaction  |
| template  | N/A| dry run with no server interaction  |

Attached are the manual tests I did in my cluster. 
[dryRunTests.txt](https://github.com/helm/helm/files/10428669/dryRunTests.txt)

QuentinN42 commented:
Oh nice work !

You've just added `--dry-run=false` twice by commands ...
Can you confirm that it's not a dry run ?

If you have some time Thursdays at 9:30-10:00AM (Pacific Time), you can propose it for review during the weekly meeting.

QuentinN42 commented:
We also need to add some docs in the https://github.com/helm/helm-www repo. Do you have something written already ?

tapaskapadia commented:
Thanks for catching that, I fixed the table. I can confirm `--dry-run=false` is not a dry operation. Actual helm outputs of the test cases in that table are in the dryRunTests.txt. 

I have not looked into the required changes for the [helm-www](https://github.com/helm/helm-www) repo yet. Does that have to be ready to go before this is reviewed? 

I am not certain I will be able to make it Thursday, but I'll try. It would be great to see this issue resolved.

QuentinN42 commented:
I've created the MR, feel free to add some commits :)

tapaskapadia committed:
feat(helm): add ability for --dry-run to do lookup functions
When a helm command is run with the --dry-run flag, it will try to connect to the cluster
if the value is 'server' to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

joejulian commented on pkg/action/install.go:
```diff
@@ -71,7 +71,7 @@ type Install struct {
 	Force                    bool
 	CreateNamespace          bool
-	DryRun                   bool
+	DryRun                   string
```
This is a breaking api change. `DryRun` needs to remain a bool. Perhaps add a new member, `DryRunOption`. All the cli flags can work using the new member but the old member would still need to be honored if true as there are other applications out there that use this as a library.

QuentinN42 replied:
And what do you think of that ?
https://github.com/helm/helm/issues/8137#issuecomment-1383664380

tapaskapadia replied:
Understood. Let me rework it with another flag.

joejulian replied:
Shoot, not quite what I was trying to suggest. I was still suggesting to just have the one single `dryrun` flag for the cli. That flag would load its input into the new `DryRunOption` string.

The reason for the new string values is that if a downstream application is using the install action as a library, it's going to have an `Install` struct with a `DryRun` boolean already. If we change the type of `DryRun`, those downstream applications will break.

So my suggestion is that the new string element would be used for the dryrun flag. In places where DryRun is checked, call a function like:
```go
func (i *Install) DryRunCheck() string {
	if i.DryRunOptions != nil {
		return i.DryRunOptions
	}
	if i.DryRun {
		return "client"
	}
	return "none"
}
```
you could simplify all the choices here and use/return a custom type instead of string, which might make the rest of the code easier:
```
type DryRunType string

var DryRunServer = DryRunType("server")
var DryRunClient = DryRunType("client")
var DryRunNone = DryRunType("none")
```

```
if DryRunCheck() == DryRunServer { ...
```

The benefit to all this is that the user experience at the command line doesn't have to change, yet we maintain backward compatibility for all the projects that use helm as a go library.

disclaimer: These snippets are used as illustrations. Don't feel like I'm dictating anything.

joejulian replied:
Don't you love it when you reply to something before you've scrolled all the way down?

tapaskapadia committed:
feat(helm): add ability for a dry-run to evaluate lookup functions
When a helm command is run with the --dry-run-option=server flag, it will try to connect to the cluster
to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

tapaskapadia commented:
I reworked the PR to add a new option --dry-run-option as suggested by @joejulian 
The --dry-run flag should remain completely unchanged. 
The new --dry-run-option flag is mutually exclusive with --dry-run and can have the values 'none','true','false','client', or 'server'. 

Below is a table of expected interactions with the flags with this change. Attached are the actual tests performed on a cluster: [dryRunTestsRound3.txt](https://github.com/helm/helm/files/10483001/dryRunTestsRound3.txt)


| command       | Flag Value | expected behavior |
| ------------- | ------------- | ------------- |
| install  | [no flag present]  | not a dry run  |
| install  | --dry-run | dry run w/ no server interaction |
| install  | --dry-run=false  | not a dry run  |
| install  | --dry-run=true | dry run w/ no server interaction  |
| install  | --dry-run-option=false  | not a dry run  |
| install  | --dry-run-option=none  | not a dry run  |
| install  | --dry-run-option  | dry run w/ no server interaction  |
| install  | --dry-run-option=true | dry run w/ no server interaction  |
| install  | --dry-run-option=client | dry run w/ no server interaction  |
| install  | --dry-run-option=server | dry run with server interaction  |
| install  | --dry-run-option=badInput | error: dry-run-option needs to be 'none','false','true','client',or 'server'  |
| install  | --dry-run=false --dry-run-option=server | error: only can set one of those two flags|
| upgrade --install  | [no flag present]  | not a dry run  |
| upgrade --install  | --dry-run  | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=false  | not a dry run  |
| upgrade --install  | --dry-run=true  | dry run w/ no server interaction |
| upgrade --install  | --dry-run-option | dry run w/ no server interaction |
| upgrade --install  | --dry-run-option=false  | not a dry run  |
| upgrade --install  | --dry-run-option=true  | dry run w/ no server interaction |
| upgrade --install  | --dry-run-option=none  | not a dry run  |
| upgrade --install  | --dry-run-option=true | dry run w/ no server interaction  |
| upgrade --install  | --dry-run-option=client | dry run w/ no server interaction  |
| upgrade --install  | --dry-run-option=server | dry run with server interaction  |
| upgrade --install  | --dry-run-option=badInput | error: dry-run-option needs to be 'none','false','true','client',or 'server') |
|  upgrade --install   | --dry-run=false --dry-run-option=server | error: only can set one of those two flags|
| template  | N/A| dry run with no server interaction  |

tapaskapadia committed:
feat(helm): add ability for a dry-run to evaluate lookup functions
When a helm command is run with the --dry-run-option=server flag, it will try to connect to the cluster
to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

tapaskapadia committed:
feat(helm): add ability for a dry-run to evaluate lookup functions
When a helm command is run with the --dry-run-option=server flag, it will try to connect to the cluster
to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

rickardp commented:
The request was for the internal API to be non breaking. This will break the symmetry with kubectl. Wasn't the idea to deprecate the parameter-less --dry-run?

pwillis-eiq commented:
> Perhaps add a new member, DryRunOption. All the cli flags can work using the new member but the old member would still need to be honored if true as there are other applications out there that use this as a library.

Based on this comment, it sounds like the actual CLI arguments can remain as originally suggested (`--dry-run`, `--dry-run=false`, --dry-run=true`, `--dry-run=client`, `--dry-run=server`, `--dry-run=none`) with no new flags, while internally the code treats `DryRun` as a *bool*, and adding an additional struct element `DryRunOption` for the other cases (`client`, `server`, `none`).

Effectively, add in code that checks if the command-line option is `--dry-run=true` or `--dry-run=false`, and if so, update `DryRun`. But if the command line is `--dry-run=client`, `--dry-run=server`, or `--dry-run=none`, update `DryRunOption` instead. If `DryRunOption` is a string, and is not empty, then the intended functionality is the latter; if the string *is* empty, then the functionality intended is the former (the *bool*).

Does this make sense / sound like it would work?

tapaskapadia committed:
feat(helm): add ability for --dry-run to do lookup functions
When a helm command is run with the --dry-run flag, it will try to connect to the cluster
if the value is 'server' to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

tapaskapadia commented:
@rickardp That is good feedback and I liked it better with only one flag as well. How about the latest change I pushed up? Internally dryRun is still present but the cli flags bind to dryRunOption not dryRun. I believe this would maintain API compatibility but I am not certain. Is there any easy way to test it?

The behavior would be as described before:
| command       | dryRun Flag Value | expected behavior |
| ------------- | ------------- | ------------- |
| install  | [no flag present]  | not a dry run  |
| install  | --dry-run=false  | not a dry run  |
| install  | --dry-run=none  | not a dry run  |
| install  | --dry-run  | dry run w/ no server interaction  |
| install  | --dry-run=true | dry run w/ no server interaction  |
| install  | --dry-run=client | dry run w/ no server interaction  |
| install  | --dry-run=server | dry run with server interaction  |
| upgrade --install  | [no flag present]  | not a dry run  |
| upgrade --install  | --dry-run=false  | not a dry run  |
| upgrade --install  | --dry-run=none  | not a dry run  |
| upgrade --install  | --dry-run  | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=true | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=client | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=server | dry run with server interaction  |
| template  | N/A| dry run with no server interaction  |

tapaskapadia commented:
I did end up validating that the current implementation is not a breaking API change. Is there any other concerns with the implementation? It would be awesome to get it reviewed if this addresses the concerns. 

Summary of the implementation:
- --dry-run cli flag is binded to a new member DryRunOption (to handle string cli data where the valid values are "none","false","true", "client", and "server". 
- `--dry-run=server` allows for lookups to be successful. 
- Original DryRun (bool) member is still respected.


Just for reference in case others are curious here's how I tested it programmatically. (used replace in the go.mod to point to my local helm package)
``` go
func main() {
	chartPath := "/Users/tapaskapadia/Documents/Code/sandbox/goSand/echoTapas-0.1.0.tgz"
	chart, err := loader.Load(chartPath)
	if err != nil {
		panic(err)
	}
	kubeconfigPath := "/Users/tapaskapadia/.kube/config"
	releaseNamespace := "tapas"
	actionConfig := new(action.Configuration)
	if err := actionConfig.Init(kube.GetConfig(kubeconfigPath, "", releaseNamespace), releaseNamespace, os.Getenv("HELM_DRIVER"), nil); err != nil {
		panic(err)
	}
	install := action.NewInstall(actionConfig)
	install.Namespace = releaseNamespace
	install.ReleaseName = "blah"
	install.DryRun = true
	//install.DryRunOption = "server"
	out, err := install.Run(chart, nil)
	if err != nil {
		panic(err)
	}
	fmt.Println(out)
}
```

joejulian commented:
With a very simple template:
```
---
lookedup: {{ lookup "v1" "Pod" "" "" | toYaml | nindent 2 }}
```

I'm not getting any data:
```
$ /home/jjulian/dev/go/src/github.com/helm/helm/bin/helm template . --dry-run=server
---
# Source: lookup-test/templates/lookup.yaml
lookedup: 
  {}
```

```
$ /home/jjulian/dev/go/src/github.com/helm/helm/bin/helm version
[lookup-test.tar.gz](https://github.com/helm/helm/files/10524199/lookup-test.tar.gz)

version.BuildInfo{Version:"v3.11+unreleased", GitCommit:"9a0025f96388bcb45f91b2206e0e9d24b498d339", GitTreeState:"clean", GoVersion:"go1.19.5"}
```

tapaskapadia commented:
@joejulian 

> 

The `--dry-run` flag is not added to the `template` function. This PR currently only modifies the existing commands that have the `dry-run` flag - upgrade / install. Can you check if the the released version of helm is outputting the data when you do `helm install --dry-run --debug` as that is strange behavior which I was not able to replicate on my side?

"""
🙈➞ git status
On branch feat/lookup-dryrun
Your branch is up to date with 'origin/feat/lookup-dryrun'.

nothing to commit, working tree clean
~/Documents/Code/helmMergeFix/helm
🙈➞ make build 
GO111MODULE=on CGO_ENABLED=0 go build  -trimpath -tags '' -ldflags '-w -s -X helm.sh/helm/v3/internal/version.metadata=unreleased -X helm.sh/helm/v3/internal/version.gitCommit=9a0025f96388bcb45f91b2206e0e9d24b498d339 -X helm.sh/helm/v3/internal/version.gitTreeState=clean  -X helm.sh/helm/v3/pkg/lint/rules.k8sVersionMajor=1 -X helm.sh/helm/v3/pkg/lint/rules.k8sVersionMinor=26 -X helm.sh/helm/v3/pkg/chartutil.k8sVersionMajor=1 -X helm.sh/helm/v3/pkg/chartutil.k8sVersionMinor=26' -o '/Users/tapaskapadia/Documents/Code/helmMergeFix/helm/bin'/helm ./cmd/helm
~/Documents/Code/helmMergeFix/helm
🙈➞ ./bin/helm  install --generate-name --dry-run ~/Documents/Code/sandbox/goSand/helmtester/
NAME: helmtester-1674869788
LAST DEPLOYED: Fri Jan 27 19:36:28 2023
NAMESPACE: default
STATUS: pending-install
REVISION: 1
TEST SUITE: None
HOOKS:
MANIFEST:
---
# Source: echoTapas/templates/lookup.yaml
# {}

~/Documents/Code/helmMergeFix/helm
🙈➞ ./bin/helm  install --generate-name --dry-run=server ~/Documents/Code/sandbox/goSand/helmtester/
NAME: helmtester-1674869797
LAST DEPLOYED: Fri Jan 27 19:36:37 2023
NAMESPACE: default
STATUS: pending-install
REVISION: 1
TEST SUITE: None
HOOKS:
MANIFEST:
---
# Source: echoTapas/templates/lookup.yaml
# {"apiVersion":"v1","items":[{"apiVersion":"v1","kind":"Pod","metadata":{"creationTimestamp":"2023-01-25T13:05:30Z","generateName":"coredns-565d847f94-","labels":{"k8s-app":"kube-dns","pod-template-hash":"565d847f94"},"managedFields":[{"apiVersion":"v1","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:generateName":{},"f:labels":{".":{},"f:k8s-app":{},"f:pod-template-hash":{}},"f:ownerReferences":{".":{},"k:{\"uid\":\"eeada5d1-9603-460a-8aa4-0adba03cf0e5\"}":{}}},"f:spec":{"f:affinity":{".":{},"f:podAntiAffinity":{".":{},"f:preferredDuringSchedulingIgnoredDuringExecution":{}}},"f:containers":{"k:{\"name\":\"coredns\"}":{".":{},"f:args":{},"f:image":{},"f:imagePullPolicy":{},"f:livenessProbe":{".":{},"f:failureThreshold":{},"f:httpGet":{".":{},"f:path":{},"f:port":{},"f:scheme":{}},"f:initialDelaySeconds":{},"f:periodSeconds":{},"f:successThreshold":{},"f:timeoutSeconds":{}},"f:name":{},"f:ports":{".":{},"k:{\"containerPort\":53,\"protocol\":\"TCP\"}":{".":{},"f:containerPort":{},"f:name":{},"f:protocol":{}},"k:{\"containerPort\":53,\"protocol\":\"UDP\"}":{".":{},"f:containerPort":{},"f:name":{},"f:protocol":{}},"k:{\"containerPort\":9153,\"protocol\":\"TCP\"}":{".":{},"f:containerPort":{},"f:name":{},"f:protocol":{}}},"f:readinessProbe":{".":{},"f:failureThreshold":{},"f:httpGet":{".":{},"f:path":{},"f:port":{},"f:scheme":{}},"f:periodSeconds":{},"f:successThreshold":{},"f:timeoutSeconds":{}},"f:resources":{".":{},"f:limits":{".":{},"f:memory":{}},"f:requests":{".":{},"f:cpu":{},"f:memory":{}}},"f:securityContext":{".":{},"f:allowPrivilegeEscalation":{},"f:capabilities":{".":{},"f:add":{},"f:drop":{}},"f:readOnlyRootFilesystem":{}},"f:terminationMessagePath":{},"f:terminationMessagePolicy":
"""

tapaskapadia commented:
My helm chart setup for that was just the following:

``` 
# {{ lookup "v1" "Pod" "" "" | toRawJson}}
```

tapaskapadia commented:
Ah well, I see that I missed that `template` also has a `dry-run`

tapaskapadia committed:
feat(helm): add ability for --dry-run to do lookup functions
When a helm command is run with the --dry-run flag, it will try to connect to the cluster
if the value is 'server' to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

tapaskapadia commented:
Updated the code to not exclude the `template` function even though it seems a little weird for template to have it. 
 
Here's the updated expected behavior table: 
| command       | dryRun Flag Value | expected behavior |
| ------------- | ------------- | ------------- |
| install  | [no flag present]  | not a dry run  |
| install  | --dry-run=false  | not a dry run  |
| install  | --dry-run=none  | not a dry run  |
| install  | --dry-run  | dry run w/ no server interaction  |
| install  | --dry-run=true | dry run w/ no server interaction  |
| install  | --dry-run=client | dry run w/ no server interaction  |
| install  | --dry-run=server | dry run with server interaction  |
| upgrade --install  | [no flag present]  | not a dry run  |
| upgrade --install  | --dry-run=false  | not a dry run  |
| upgrade --install  | --dry-run=none  | not a dry run  |
| upgrade --install  | --dry-run  | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=true | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=client | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=server | dry run with server interaction  |
| template  | [no flag present] | dry run w/ no server interaction  |
| template  | --dry-run | dry run w/ no server interaction  |
| template  | --dry-run=client | dry run w/ no server interaction  |
| template  | --dry-run=server | dry run with server interaction  |
| template  | --dry-run=none | dry run with server interaction  |
| template  | --dry-run=false | dry run with server interaction  |

false and none are a little strange for the template function but it keeps the behavior consistent. 

Tests of each of the expected behaviors on a cluster: 
[testCommandWithTemplate.txt](https://github.com/helm/helm/files/10524349/testCommandWithTemplate.txt)

joejulian reviewed (approved):
It all works for me. Nice work!

tapaskapadia committed:
feat(helm): add ability for --dry-run to do lookup functions
When a helm command is run with the --dry-run flag, it will try to connect to the cluster
if the value is 'server' to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

tapaskapadia commented:
Thanks for taking the time to review it!

QuentinN42 commented:
Nice work !
Can't wait to test it as soon as it's released :smiley:

crenshaw-dev commented:
I agree that the issue thread seemed to indicate core maintainers wouldn't support a change to `helm template` behavior. But I'm excited to see `helm template` support in this PR, because it lowers the threshold for Argo CD to make use of a server-side dry run. We'll still need to find a way to make this feature available in a secure way. But we'll be a step closer. 🙂

QuentinN42 commented:
I think the `template` function is really intended to stay offline command.
You don't want to template your productions variable in your chart while creating it.
But if you want to check the chart you can use `helm upgrade --dry-run` or use the [helm-diff plugin](https://github.com/databus23/helm-diff).

tapaskapadia commented:
If the `template` function is going to be a hard sell due to security concerns for the core maintainers, let me know if I should change it back to exclude the template function.

tapaskapadia committed:
feat(helm): add ability for --dry-run to do lookup functions

When a helm command is run with the --dry-run flag, it will try to connect to the cluster
if the value is 'server' to be able to render lookup functions.
Closes helm#8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

MichaelMorrisEst commented on cmd/helm/install.go:
```diff
@@ -154,7 +154,8 @@ func newInstallCmd(cfg *action.Configuration, out io.Writer) *cobra.Command {
 func addInstallFlags(cmd *cobra.Command, f *pflag.FlagSet, client *action.Install, valueOpts *values.Options) {
 	f.BoolVar(&client.CreateNamespace, "create-namespace", false, "create the release namespace if not present")
-	f.BoolVar(&client.DryRun, "dry-run", false, "simulate an install")
+	f.StringVar(&client.DryRunOption, "dry-run", "none", "simulate an install. If --dry-run is set with no option being specified or as 'client', it will not attempt cluster connections. Setting option as 'server' allows attempting cluster connections.")
```
perhaps worth clarifying the cluster connection will only be used for read type operations, nothing will be changed on the cluster

MichaelMorrisEst commented:
Ran manual tests of the various dry-run flag options and all looks good with the behaviour of each as documented. Looking forward to this being merged as it will potentially help to significantly reduce our upgrade times using helmfile

mattfarina commented:
@tapaskapadia this PR looks fairly good. Can you rebase it please.

tapaskapadia committed:
feat(helm): add ability for --dry-run to do lookup functions

When a helm command is run with the --dry-run flag, it will try to connect to the cluster
if the value is 'server' to be able to render lookup functions.
Closes helm#8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

tapaskapadia commented:
@mattfarina I updated the PR.

Upon sanity testing when updating the PR, I did also update a couple things.

1. Made the --dry-run flag usage description more explicitly state that if you are setting the `--dry-run` flag you must use `=` like `--dry-run=server`. `--dry-run server` is not valid and would result in using the set `NoOptDefVal`. `NoOptDefVal` is being utilized to allow for just setting `--dry-run` with no value as is done currently. https://github.com/spf13/pflag/issues/134 describes this expected behavior for `NoOptDefVal`.
2. Fix unhandled case for `helm template` by using `unchanged` as the default for `DryRunOption` and handle that differently for template vs install / upgrade commands.

Here's an the expected behavior table and a text file with my matching test results [test_results_03-08-23.txt](https://github.com/helm/helm/files/11184505/test_03-08-23.txt):
| command       | dryRun Flag Value | expected behavior |
| ------------- | ------------- | ------------- |
| install  | --dry-run  | dry run w/ no server interaction  |
| install  | --dry-run=true | dry run w/ no server interaction  |
| install  | --dry-run=client | dry run w/ no server interaction  |
| install  | --dry-run=server | dry run with server interaction  |
| install  | --dry-run=badinput  | error for invalid option |
| install  | --dry-run=none  | not a dry run  |
| install  | --dry-run=false  | not a dry run  |
| install  | [no flag present]  | not a dry run  |
| upgrade --install  | --dry-run  | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=true | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=client | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=server | dry run with server interaction  |
| upgrade --install  | --dry-run=badinput | error for invalid option  |
| upgrade --install  | --dry-run=none  | not a dry run  |
| upgrade --install  | --dry-run=false  | not a dry run  |
| upgrade --install  | [no flag present]  | not a dry run  |
| template  | --dry-run | dry run w/ no server interaction  |
| template  | --dry-run=client | dry run w/ no server interaction  |
| template  | --dry-run=server | dry run with server interaction  |
| template  | --dry-run=true | dry run w/ no server interaction  |
| template  | --dry-run=false | dry run with server interaction  |
| template  | --dry-run=badinput | error for invalid option |
| template  | --dry-run=none | dry run with server interaction  |
| template  | [no flag present] | dry run w/ no server interaction  |

MichaelMorrisEst commented:
Looks good, only question I'd have is if  "--dry-run=false" and "--dry-run=none" should be considered invalid for template command as they dont really make sense.
If I am not mistaken setting "--dry-run=false" currently has no effect for the template command as the actual value used is hard coded to true (template.go line 82)

@mattfarina It would be great if you could review the latest commit. I would very much like to get this in 3.12.0 if possible

mattfarina commented on cmd/helm/template.go:
```diff
@@ -79,6 +79,9 @@ func newTemplateCmd(cfg *action.Configuration, out io.Writer) *cobra.Command {
 			}
 			client.SetRegistryClient(registryClient)
 
+			if client.DryRunOption == "unchanged" {
+				client.DryRunOption = "true"
+			}
```
If someone passes in `--dry-run=server` or `--dry-run=false` then this code is skipped and Helm would interact with the server, if I read RunWithContext change right. This has an impact on the security model of Helm where we intentionally don't want to make calls to the server for `helm template`. If we opted in to those calls it would need to be explicit and not related to dry run. This model on `helm template` came out of past security audits from 3rd parties.

Can you please make changes so we can keep the security model on `helm template`.

tapaskapadia replied:
@mattfarina To make sure I am fully understanding the changes requested here, is the desire to have `helm template` behavior not change at all and continue to have lookups fail regardless of what is passed in for the dryRun flag? I was under the impression it was fair game but if that goes against the security model, I'm more than happy to change it.  

Using the expected behavior table format I've posted in the thread, is this ideal state for helm template?

command | dryRun Flag Value | expected behavior
-- | -- | --
template | --dry-run | dry run w/ no server interaction
template | --dry-run=client | dry run w/ no server interaction
template | --dry-run=server | dry run w/ no server interaction (potentially throw invalid option error)
template | --dry-run=true | dry run w/ no server interaction
template | --dry-run=false | dry run w/ no server interaction  (potentially throw invalid option error) 
template | --dry-run=badinput | error for invalid option 
template | --dry-run=none | dry run w / no server interaction (potentially throw invalid option error)
template | [no flag present] | dry run w/ no server interaction

mattfarina replied:
@tapaskapadia I went back to my notes and I think your original implementation works where it allows some connection to the cluster as an opt-in. I went back to your previous table. Sorry I didn't clearly remember the previous conversation when I made this.

tapaskapadia commented on cmd/helm/install.go:
```diff
@@ -160,7 +163,8 @@ func newInstallCmd(cfg *action.Configuration, out io.Writer) *cobra.Command {
 func addInstallFlags(cmd *cobra.Command, f *pflag.FlagSet, client *action.Install, valueOpts *values.Options) {
 	f.BoolVar(&client.CreateNamespace, "create-namespace", false, "create the release namespace if not present")
-	f.BoolVar(&client.DryRun, "dry-run", false, "simulate an install")
+	f.StringVar(&client.DryRunOption, "dry-run", "unchanged", "simulate an install. If --dry-run is set with no option being specified or as '--dry-run=client', it will not attempt cluster connections. Setting '--dry-run=server' allows attempting cluster connections.")
+	f.Lookup("dry-run").NoOptDefVal = "client"
```
Great call out. I'll change it to `""` so end users are not exposed to this internal default value.

mattfarina replied:
@tapaskapadia I think this is my one nit I would like to see changed before approving.

tapaskapadia replied:
@mattfarina I just pushed up the changes to make the default value to be "" which make the help documentation much clearer. Here's how the help documentation reads now:
```
--dry-run string[="client"]                  simulate an install. If --dry-run is set with no option being specified or as '--dry-run=client', it will not attempt cluster connections. Setting '--dry-run=server' allows attempting cluster connections.

```
Here's the expected behavior table (nothing has changed) with my matching test results including the latest changes: [test_03-30-23.txt](https://github.com/helm/helm/files/11363626/test_03-30-23.txt)

| command       | dryRun Flag Value | expected behavior |
| ------------- | ------------- | ------------- |
| install  | --dry-run  | dry run w/ no server interaction  |
| install  | --dry-run=true | dry run w/ no server interaction  |
| install  | --dry-run=client | dry run w/ no server interaction  |
| install  | --dry-run=server | dry run with server interaction  |
| install  | --dry-run=badinput  | error for invalid option |
| install  | --dry-run=none  | not a dry run  |
| install  | --dry-run=false  | not a dry run  |
| install  | [no flag present]  | not a dry run  |
| upgrade --install  | --dry-run  | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=true | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=client | dry run w/ no server interaction  |
| upgrade --install  | --dry-run=server | dry run with server interaction  |
| upgrade --install  | --dry-run=badinput | error for invalid option  |
| upgrade --install  | --dry-run=none  | not a dry run  |
| upgrade --install  | --dry-run=false  | not a dry run  |
| upgrade --install  | [no flag present]  | not a dry run  |
| template  | --dry-run | dry run w/ no server interaction  |
| template  | --dry-run=client | dry run w/ no server interaction  |
| template  | --dry-run=server | dry run with server interaction  |
| template  | --dry-run=true | dry run w/ no server interaction  |
| template  | --dry-run=false | dry run with server interaction  |
| template  | --dry-run=badinput | error for invalid option |
| template  | --dry-run=none | dry run with server interaction  |
| template  | [no flag present] | dry run w/ no server interaction  |

crenshaw-dev replied:
Is there any further action needed here? I'm happy to help in any way possible.

philomory replied:
@mattfarina @tapaskapadia Are there any additional changes required before this can be merged? It seems like all of the requested changes have been addressed.

tapaskapadia replied:
@philomory I believe I have implemented all the requested changes. If there’s anything else needed, I happy to make changes.

tapaskapadia commented on cmd/helm/install.go:
```diff
@@ -308,3 +317,19 @@ func compInstall(args []string, toComplete string, client *action.Install) ([]st
 	return nil, cobra.ShellCompDirectiveNoFileComp
 }
+
+func validateDryRunOptionFlag(dryRunOptionFlagValue string) error {
```
validateDryRunOptionFlag function does run for template. cmd/helm/template.go calls the function runInstall [here](https://github.com/helm/helm/blob/4b7248e361e60b901c972d671c78a2fb38d99da2/cmd/helm/template.go#L91) and then runInstall function calls the validateDryRunOptionFlag [here](https://github.com/helm/helm/blob/4b7248e361e60b901c972d671c78a2fb38d99da2/cmd/helm/install.go#L276). I can move it around if desired but I remember seeing other validation in runInstall so I thought it fit well there.

tapaskapadia committed:
feat(helm): add ability for --dry-run to do lookup functions
When a helm command is run with the --dry-run flag, it will try to connect to the cluster
to be able to render lookup functions.
Closes #8137

Signed-off-by: Tapas Kapadia <tapaskapadia10@gmail.com>

dudicoco commented:
@tapaskapadia thank you for this important contribution!

pwillis-eiq commented:
Thank you @tapaskapadia, @rickardp, @joejulian, @QuentinN42, @mattfarina, and everyone else who helped push this through!

This feature change took 2 years, 4 months and 20 days to merge! But it was finally completed - by a bunch of strangers on the internet, all working together, with no monetary incentive, contractual agreement, or under any hierarchical organization chart. Because of all this, a bunch of other strangers will now benefit - whether it's just getting their day jobs done, or working on the tools they care about. Open Source lifts all boats ❤️ ⛵ ❤️
